sharedbuffers/mapped_struct.py

```diff
@@ -333,14 +335,231 @@ class mapped_dict(dict):
 
     @classmethod
     def pack_into(cls, obj, buf, offs, idmap = None, implicit_offs = 0):
-        keys = list(obj)
-        values = [obj[k] for k in keys]
-        return mapped_list.pack_into([keys, values], buf, offs, idmap, implicit_offs)
+        return proxied_dict.pack_into(obj, buf, offs, idmap, implicit_offs)
+
+    @classmethod
+    def unpack_from(cls, buf, offs, idmap = None):
+        proxy = proxied_dict.unpack_from(buf, offs, idmap)
+        return proxy.copy()
+
+
+@cython.ccall
+@cython.locals(code=cython.ulong, nbits=cython.ulong)
+def _hash_rotl(code, nbits):
+    return (code << nbits) | (code >> (32 - nbits))
+
+
+@cython.ccall
+@cython.locals(code1=cython.ulong, code2=cython.ulong)
+def _mix_hash(code1, code2):
+    return _hash_rotl(code1, 5) ^ code2
+
+
+_TUPLE_SEED = 4110121799
+_FSET_SEED  = 3377403095
+
+@cython.ccall
+@cython.locals(hval=cython.ulong)
+def _stable_hash(key):
+    if isinstance(key, basestring):
+        hval = xxhash.xxh64(safe_utf8(key)).intdigest()
+    elif isinstance(key, int):
+        hval = key
+    elif isinstance(key, float):
+        trunc_key = int(key)
+        if trunc_key == key:
+            hval = trunc_key
+        else:
+            mant, expo = math.frexp(key)
+            hval = _mix_hash(expo, int(mant * 0xffffffffffff))
+    elif isinstance(key, (tuple, frozenset, proxied_tuple)):
+        if isinstance(key, frozenset):
+            hval = _FSET_SEED
+        else:
+            hval = _TUPLE_SEED
+
+        for value in key:
+            hval = _mix_hash(hval, _stable_hash(value))
+    else:
+        raise TypeError("unhashable type: %s" % type(key).__name__)
+
+    return hval
+
+
+def _enum_keys(obj):
+    idx = 0
+    for key in obj.iterkeys():
+        yield key, idx
+        idx += 1
+
+
+class BufferIO(object):
+    """
+    Gimmick object. Its whole purpose is to implement
+    the protocol expected by the IdMapper classes.
+    """
+
+    def __init__(self, buf, offs):
+        self.buf = buf
+        self.offs = offs
+        self.pos = 0
+
+    def write(self, value):
+        size = len(value)
+        offs = self.offs + self.pos
+        self.buf[offs:offs + size] = value
+        self.pos += size
+        return size
+
+    def tell(self):
+        return self.pos
+
+    def flush(self):
+        pass
+
+    def seek(self, pos):
+        self.pos = pos
+
+
+@cython.cclass
+class proxied_dict(object):
+
+    HEADER_PACKER = struct.Struct('=Q')   # Offset into value list.
+
+    cython.declare(objmapper=object, vlist=proxied_list)
+
+    def __init__(self, objmapper, vlist):
+        self.objmapper = objmapper
+        self.vlist = vlist
+
+    @classmethod
+    def pack_into(cls, obj, buf, offs, idmap = None, implicit_offs = 0):
+        packer = cls.HEADER_PACKER
+        ipos = offs
+        offs += packer.size
+        iobuf = BufferIO(buf, offs)
+        mapper_size = ObjectIdMapper.build(_enum_keys(obj), iobuf, map_file=False)
```

Comment:
Maybe you should implement the `map_file=False` option in all id mapper classes to have a consistent convention

And I'd rather call it `return_mapper=False`